### ISSUE TITLE
Highlighting identifiers w/"‾" in them

### ISSUE THREAD

**Issue by jasonhemann** [greghendershott/racket-mode#534]
On my system, I notice that for any identifiers w/ that symbol in them, the highlighting doesn't catch "‾" or anything after it.

```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
		 (20200810 1513)
		 "Racket editing, REPL, and more"
		 ((emacs
		   (25 1))
		  (faceup
		   (0 0 2))
		  (pos-tip
		   (20191127 1028)))
		 nil nil "/Users/jhemann/.emacs.d/elpa/racket-mode-20200810.1513"
		 ((:url . "https://www.racket-mode.com/")
		  (:maintainer "Greg Hendershott")
		  (:authors
		   ("Greg Hendershott"))
		  (:commit . "c55fd70c5e9a371f262486f7d5d8837481388b95"))
		 nil))
((emacs-version "27.2")
 (system-type darwin)
 (x-gtk-use-system-tooltips UNDEFINED)
 (major-mode racket-repl-mode)
 (racket--el-source-dir "/Users/jhemann/.emacs.d/elpa/racket-mode-20200810.1513/")
 (racket--rkt-source-dir "/Users/jhemann/.emacs.d/elpa/racket-mode-20200810.1513/racket/")
 (racket-program "racket")
 (racket-command-timeout 10)
 (racket-xp-after-change-refresh-delay 1)
 (racket-xp-highlight-unused-regexp "^[^_]")
 (racket-repl-buffer-name-function nil)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-images-system-viewer "open")
 (racket-images-system-viewer "open")
 (racket-use-repl-submit-predicate nil)
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal)))
 (racket-show-functions
  (racket-show-pseudo-tooltip)))
(enabled-minor-modes
 (async-bytecomp-package-mode)
 (auto-complete-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (blink-cursor-mode)
 (cl-old-struct-compat-mode)
 (column-number-mode)
 (electric-indent-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-auto-complete-mode)
 (global-eldoc-mode)
 (global-flycheck-mode)
 (global-font-lock-mode)
 (global-nlinum-mode)
 (helm-ff-cache-mode)
 (helm-mode)
 (line-number-mode)
 (mac-auto-operator-composition-mode)
 (mac-mouse-wheel-mode)
 (menu-bar-mode)
 (nlinum-mode)
 (paredit-mode)
 (savehist-mode)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (straight-package-neutering-mode)
 (straight-use-package-mode)
 (tooltip-mode)
 (transient-mark-mode))
(disabled-minor-modes
 (abbrev-mode)
 (archive-subfile-mode)
 (artbollocks-mode)
 (auto-fill-function)
 (auto-save-visited-mode)
 (buffer-face-mode)
 (buffer-read-only)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (defining-kbd-macro)
 (dired-hide-details-mode)
 (eclim-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (flycheck-mode)
 (flymake-mode)
 (flyspell-mode)
 (flyspell-popup-auto-correct-mode)
 (global-eclim-mode)
 (global-hl-line-mode)
 (global-linum-mode)
 (global-prettify-symbols-mode)
 (global-visual-line-mode)
 (gradle-mode)
 (gud-tooltip-mode)
 (helm--minor-mode)
 (helm--remap-mouse-mode)
 (helm-autoresize-mode)
 (helm-display-line-numbers-mode)
 (helm-ff--delete-async-modeline-mode)
 (helm-migemo-mode)
 (helm-popup-tip-mode)
 (hl-line-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (ido-everywhere)
 (isearch-mode)
 (ispell-minor-mode)
 (jit-lock-debug-mode)
 (linum-mode)
 (mac-auto-ascii-mode)
 (mac-font-panel-mode)
 (mc-hide-unmatched-lines-mode)
 (midnight-mode)
 (mouse-wheel-mode)
 (multiple-cursors-mode)
 (next-error-follow-minor-mode)
 (org-cdlatex-mode)
 (org-src-mode)
 (orgstruct-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (prettify-symbols-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (rectangle-mark-mode)
 (rectangular-region-mode)
 (ruler-mode)
 (sh-electric-here-document-mode)
 (size-indication-mode)
 (straight-live-modifications-mode)
 (straight-symlink-emulation-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (table-fixed-width-mode)
 (table-mode-indicator)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (tool-bar-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (window-divider-mode)
 (writegood-mode)
 (xref-etags-mode)
 (yas-global-mode)
 (yas-minor-mode))
```

**Comment by greghendershott** [comment]
Thanks for the report; I can reproduce this.

It seems to be because Emacs by default classifies that character as "punctuation".

Even e.g. in a `fundamental-mode` buffer, `M-x describe-char` for that says (for me):

```
             position: 33 of 54 (59%), column: 9
            character: ‾ (displayed as ‾) (codepoint 8254, #o20076, #x203e)
    preferred charset: unicode (Unicode (ISO10646))
code point in charset: 0x203E
               script: symbol
               syntax: . 	which means: punctuation
             category: .:Base, >:Not at bol, c:Chinese, r:Roman
             to input: type "C-x 8 RET 203e" or "C-x 8 RET OVERLINE"
          buffer code: #xE2 #x80 #xBE
            file code: #xE2 #x80 #xBE (encoded by coding system utf-8-unix)
              display: by this font (glyph code)
    xft:-GNU -FreeSans-normal-normal-normal-*-24-*-*-*-*-0-iso10646-1 (#xC9B)

Character code properties: customize what to show
  name: OVERLINE
  old-name: SPACING OVERSCORE
  general-category: Po (Punctuation, Other)
  decomposition: (compat 32 773) (compat ' ' '̅')
```

And Racket Mode says that identifiers consist of one or more characters classified as "symbol" or "word".

---

In addition to what you report, I see that with `racket-xp-mode` enabled: When point is on the symbol, the entire identifier is highlighted... but the `‾` isn't as wide, making that whole segment "jump" horizontally. (This might be related to the specific fonts I'm using.)

---

Although I can think of a quick hack or two, I may need to give this a longer think for a more-general solution.

**Comment by greghendershott** [comment]
Some of the char-syntax and font-lock stuff is among the oldest parts of Racket Mode. Meaning, in hindsight, I could probably do better, today.

In this case, I notice there are quite a few regexps doing things like `(1+ (or (syntax word) (syntax symbol) space))`, to find identifier symbols. Which seemed reasonable at the time, and works for most "common" things.

But it wouldn't match an identifier using `‾`. And of course Racket is _extraordinarily_ flexible about identifier symbols. ["A sequence that does not start with a delimiter or # is parsed as either a symbol, a number....  A successful number or extflonum parse takes precedence over a symbol parse."](https://docs.racket-lang.org/reference/reader.html#(part._parse-symbol))

At some later point I did go back and try to implement the complete reader grammar for numbers. Assuming I did that correctly, then probably I should rely on that now -- i.e., "if it didn't parse as a number, it's a symbol". As a first approximation, that would probably fix this. It might surface some new problems. Fortunately I have some _fairly_ extensive tests for font-lock to help.

So I'll put that on my to-do list to try.
